config/chassis_modules.py

```diff
@@ -17,6 +21,74 @@ def modules():
     """Configure chassis modules"""
     pass
 
+
+def get_config_module_state(db, chassis_module_name):
+    config_db = db.cfgdb
+    fvs = config_db.get_entry('CHASSIS_MODULE', chassis_module_name)
+    if not fvs:
+        return 'up'
+    else:
+        return fvs['admin_status']
+
+
+#
+# Name: get_config_module_state_timeout
+# return: True: timeout, False: not timeout
+#
+def get_config_module_state_timeout(ctx, db, chassis_module_name, state):
+    counter = 0
+    while get_config_module_state(db, chassis_module_name) != state:
+        time.sleep(1)
+        counter += 1
+        if counter >= TIMEOUT_SECS:
+            ctx.fail("get_config_module_state {} timeout".format(chassis_module_name))
+            return True
+    return False
+
+
+def get_asic_list_from_db(chassisdb, chassis_module_name):
+    asic_list = []
+    asics_keys_list = chassisdb.keys("CHASSIS_STATE_DB", "CHASSIS_FABRIC_ASIC_TABLE*")
+    for asic_key in asics_keys_list:
+        name = chassisdb.get("CHASSIS_STATE_DB", asic_key, "name")
+        if name == chassis_module_name:
+            asic_id = int(re.search(r"(\d+)$", asic_key).group())
+            asic_list.append(asic_id)
+    return asic_list
+
+
+#
+# Syntax: fabric_module_set_admin_status <chassis_module_name> <'up'/'down'>
+#
+def fabric_module_set_admin_status(db, chassis_module_name, state):
+    chassisdb = db.db
+    chassisdb.connect("CHASSIS_STATE_DB")
+    asic_list = get_asic_list_from_db(chassisdb, chassis_module_name)
+
+    if len(asic_list) == 0:
+        return
+
+    if state == "down":
+        for asic in asic_list:
+            click.echo("Stop swss@{} and syncd@{} services".format(asic, asic))
+            clicommon.run_command('sudo systemctl stop swss@{}.service'.format(asic))
+
+        # wait for service is down
+        time.sleep(2)
+        click.echo("Delete related CAHSSIS_FABRIC_ASIC_TABLE entries")
+
+        for asic in asic_list:
+            chassisdb.delete("CHASSIS_STATE_DB", "CHASSIS_FABRIC_ASIC_TABLE|asic" + str(asic))
+
+        for asic in asic_list:
```

Comment:
This entry needs to be deleted.  Otherwise,  if user set admin up,  this entry will trigger the syncd to initialize Fabric too early before the hardware is ready for it. This cause crash.

Reply:
My comment was - why we are restarting the swss services for the NPUs on the fabric card which is down. (line# 86)

Reply:
Swss services are started for all fabric asics after bootup. They are staying in wait state till the fabric card is online.  Then that service changes to start state which will bring up swss container and finally changes back to wait state.
This is the services dump from one chassis where slot 3 and 6 have fabric cards.
root      764593       1  0 May08 ?        00:00:00 /bin/bash /usr/local/bin/swss.sh wait 7
root      765252       1  0 May08 ?        00:00:00 /bin/bash /usr/local/bin/swss.sh wait 0
root      765255       1  0 May08 ?        00:00:00 /bin/bash /usr/local/bin/swss.sh wait 1
root      765261       1  0 May08 ?        00:00:00 /bin/bash /usr/local/bin/swss.sh wait 12
root      765263       1  0 May08 ?        00:00:00 /bin/bash /usr/local/bin/swss.sh wait 13
root      765266       1  0 May08 ?        00:00:00 /bin/bash /usr/local/bin/swss.sh wait 14
root      765267       1  0 May08 ?        00:00:00 /bin/bash /usr/local/bin/swss.sh wait 15
root      765272       1  0 May08 ?        00:00:00 /bin/bash /usr/local/bin/swss.sh wait 2
root      765274       1  0 May08 ?        00:00:00 /bin/bash /usr/local/bin/swss.sh wait 3
root      765289       1  0 May08 ?        00:00:00 /bin/bash /usr/local/bin/swss.sh wait 6
root      765292       1  0 May08 ?        00:00:00 /bin/bash /usr/local/bin/swss.sh wait 8
root      765295       1  0 May08 ?        00:00:00 /bin/bash /usr/local/bin/swss.sh wait 9
root     1183800       1  0 May08 ?        00:00:00 /bin/bash /usr/local/bin/swss.sh wait 10
root     1183981       1  0 May08 ?        00:00:00 /bin/bash /usr/local/bin/swss.sh wait 5
root     1183982       1  0 May08 ?        00:00:00 /bin/bash /usr/local/bin/swss.sh wait 11
root     1183999       1  0 May08 ?        00:00:00 /bin/bash /usr/local/bin/swss.sh wait 4

Reply:
@mlok-nokia Another related question - I assume since the module state is set to down earlier, chassisd will not populate the "CHASSIS_FABRIC_ASIC_TABLE|asic" again for this asic's after the module update interval CHASSIS_INFO_UPDATE_PERIOD_SECS - is that right?

Reply:
Correct.  It will be updated only if the MODULE admin_status is UP (or module admin_status default value which is UP )